pkg/controlplane/control/manager.go

```diff
@@ -193,10 +233,87 @@ func (m *Manager) AddImport(ctx context.Context, imp *v1alpha1.Import) (err erro
 	return m.addImportService(ctx, imp, userService)
 }
 
+// deleteClDataplane deletes pod to ipToPod list.
+func (m *Manager) deleteEndpoint(esID types.NamespacedName) {
+	if strings.Contains(esID.Name, dpapp.Name) {
+		m.logger.Warnf("Detected cl-dataplane(%s) endpointslice delete in namespace: %s.", esID.Name, esID.Namespace)
+	}
+}
+
+// addPod adds or updates pod to ipToPod and podList.
+func (m *Manager) addEndpoint(ctx context.Context, eslice *discv1.EndpointSlice) {
+	m.podLock.Lock()
+	defer m.podLock.Unlock()
+
+	if eslice.Labels[labelServiceName] != dpapp.Name {
+		return
+	}
+	m.logger.Infof("Detected cl-dataplane(%s) endpointslice add/update in namespace: %s", eslice.Name, eslice.Namespace)
+	m.endpoints[eslice.Namespace] = eslice.Endpoints
+	// Retrieve endpointslices managed by clusterlink in the namespace and update them
+	labelSelector := labels.Set(map[string]string{
+		labelESManagedBy: appName,
+	})
+	listOpts := []client.ListOption{
+		client.InNamespace(eslice.Namespace),
+		client.MatchingLabels(labelSelector),
+	}
+
+	endpointSliceList := &discv1.EndpointSliceList{}
+	err := m.client.List(ctx, endpointSliceList, listOpts...)
+	if err != nil {
+		m.logger.Errorf("Failed to list endpointslices managed by %s", appName)
+		return
+	}
+
+	for i := range endpointSliceList.Items {
+		es := endpointSliceList.Items[i]
+		m.logger.Infof("Updating endpointslice %s", es.Name)
+		es.Endpoints = eslice.Endpoints
+		err := m.client.Update(ctx, &es)
+		if err != nil {
+			m.logger.Errorf("Failed to update endpointslice %s: %v.", appName, err)
+		}
+	}
+}
+
 // DeleteImport removes the listening socket of a previously imported service.
 func (m *Manager) DeleteImport(ctx context.Context, name types.NamespacedName) error {
 	m.logger.Infof("Deleting import '%s/%s'.", name.Namespace, name.Name)
 
+	defer m.ports.Release(name)
```

Comment:
I think the entire handling you added should be moved to the end of the function, after the original call to `m.ports.Release`, and hence drop the defer you added here.
If it's a merge-type import, the respective service should not have the label which indicates managed by clusterlink, so the service should not be deleted.

Reply:
By doing that, won't we be returning an unnecessary `conflicingServiceError` from `deleteImportService` by doing so? I think I will anyway move the section to `deleteImportEndpointSlice`

Reply:
The `conflictingServiceError` in the case of `DeleteImport` is actually a warning, as it is wrapped by `reconcile.TerminalError`, which means "don't retry".
So you will see an error in the log, and that's it.
There's no definite way to tell if this warning is justified, since the import object is already deleted, so you cannot tell if it was a merge-type import.

Reply:
Fixed it